### ISSUE TITLE
[BUG] System logs skipped on startup

### ISSUE THREAD

**Issue by wuminqi** [iamolegga/nestjs-pino#387]
Hi,

I am using pino with configuration as illustrated below:

```typescript
// app.module.ts
import { Module } from '@nestjs/common';
import { LoggerModule } from 'nestjs-pino';
import { AppController } from './app.controller';
import { AppService } from './app.service';

import * as pino from 'pino';

const logger = pino({
  level: process.env.NODE_ENV !== 'prod' ? 'debug' : 'info',
},pino.destination({
  minLength: 4096,
  sync: false,
}));

setInterval(function() {
  logger.flush();
}, 5000).unref();

@Module({
  imports: [
    LoggerModule.forRoot({
      pinoHttp: {
        logger,
        autoLogging: true,
        serializers: {
          err: pino.stdSerializers.err,
          req: r => {
            return {
              id: r.id,
              method: r.method,
              url: r.url,
            };
          },
          res: pino.stdSerializers.res,
        }
      },
    }),
  ],
  controllers: [AppController],
  providers: [AppService],
})
export class AppModule {}
```

and
```typescript
// main.ts
import { NestFactory } from '@nestjs/core';
import { Logger } from 'nestjs-pino';
import { AppModule } from './app.module';

async function bootstrap() {
  const app = await NestFactory.create(AppModule, { logger: false });
  app.useLogger(app.get(Logger));
  await app.listen(3000);
}
bootstrap();
```

But the logging output missed the system logging part.

```bash
> nest start --watch
[10:11:49 PM] Starting compilation in watch mode...

[10:11:53 PM] Found 0 errors. Watching for file changes.

{"level":30,"time":1604671913986,"pid":28435,"hostname":"2070iMac","context":"RoutesResolver","msg":"AppController {}:"}
{"level":30,"time":1604671913987,"pid":28435,"hostname":"2070iMac","context":"RouterExplorer","msg":"Mapped {, GET} route"}
{"level":30,"time":1604671913989,"pid":28435,"hostname":"2070iMac","context":"NestApplication","msg":"Nest application successfully started"}
```

It is expected to output as many as system logging output, which has logging form  modules like NestFactory and InstanceLoader. Below is the output from the default logger.

```bash
[10:33:59 PM] Starting compilation in watch mode...

[10:34:00 PM] Found 0 errors. Watching for file changes.

[Nest] 30928   - 11/06/2020, 10:34:01 PM   [NestFactory] Starting Nest application...
[Nest] 30928   - 11/06/2020, 10:34:01 PM   [InstanceLoader] LoggerModule dependencies initialized +15ms
[Nest] 30928   - 11/06/2020, 10:34:01 PM   [InstanceLoader] LoggerCoreModule dependencies initialized +1ms
[Nest] 30928   - 11/06/2020, 10:34:01 PM   [InstanceLoader] AppModule dependencies initialized +0ms
[Nest] 30928   - 11/06/2020, 10:34:01 PM   [RoutesResolver] AppController {}: +6ms
[Nest] 30928   - 11/06/2020, 10:34:01 PM   [RouterExplorer] Mapped {, GET} route +1ms
[Nest] 30928   - 11/06/2020, 10:34:01 PM   [NestApplication] Nest application successfully started +4ms
```

Here are some sample code to reproduce this issue: https://github.com/wuminqi/example-pino-nestjs

Can you kindly offer a help on this?

Thanks.

**Comment by iamolegga** [comment]
Hello, thanks for reporting, this is know issue, I've answered it before [here](https://github.com/iamolegga/nestjs-pino/issues/125#issuecomment-714688800).

I'll convert this to issue and keep it open, but not promising to solve this in nearest future, because that's require some refactoring, but from my point of view that's not so critical for most users. Some related discussions could be found here https://github.com/iamolegga/nestjs-pino/issues/353

**Comment by wuminqi** [comment]
Hi, thanks for your kindly reply. 

I realize this issue as I misconfigured TypeORM and all the error logs from [TypeOrmModule] is swallowed. I have no idea what exception it is. 

I updated the sample repo by adding typeorm and some invalid configurations. 

Using nestjs-pino, the startup output
```bash
[8:57:07 AM] Starting compilation in watch mode...

[8:57:09 AM] Found 0 errors. Watching for file changes.





```

Using default logger, the startup output:
```bash
[8:50:58 AM] Starting compilation in watch mode...

[8:51:00 AM] Found 0 errors. Watching for file changes.

[Nest] 32827   - 11/07/2020, 8:51:00 AM   [NestFactory] Starting Nest application...
[Nest] 32827   - 11/07/2020, 8:51:00 AM   [InstanceLoader] LoggerModule dependencies initialized +48ms
[Nest] 32827   - 11/07/2020, 8:51:00 AM   [InstanceLoader] TypeOrmModule dependencies initialized +0ms
[Nest] 32827   - 11/07/2020, 8:51:00 AM   [InstanceLoader] LoggerCoreModule dependencies initialized +1ms
[Nest] 32827   - 11/07/2020, 8:51:00 AM   [InstanceLoader] AppModule dependencies initialized +0ms
[Nest] 32827   - 11/07/2020, 8:51:10 AM   [TypeOrmModule] Unable to connect to the database. Retrying (1)... +10000ms
Error: Handshake inactivity timeout
    at Handshake.<anonymous> (/Users/minqi/workspace/dirtySpace/aa/example-pino-nestjs/node_modules/mysql/lib/protocol/Protocol.js:160:17)
    at Handshake.emit (events.js:311:20)
    at Handshake._onTimeout (/Users/minqi/workspace/dirtySpace/aa/example-pino-nestjs/node_modules/mysql/lib/protocol/sequences/Sequence.js:124:8)
    at Timer._onTimeout (/Users/minqi/workspace/dirtySpace/aa/example-pino-nestjs/node_modules/mysql/lib/protocol/Timer.js:32:23)
    at listOnTimeout (internal/timers.js:549:17)
    at processTimers (internal/timers.js:492:7)
    --------------------
    at Protocol._enqueue (/Users/minqi/workspace/dirtySpace/aa/example-pino-nestjs/node_modules/mysql/lib/protocol/Protocol.js:144:48)
    at Protocol.handshake (/Users/minqi/workspace/dirtySpace/aa/example-pino-nestjs/node_modules/mysql/lib/protocol/Protocol.js:51:23)
```

**Comment by AliYusuf95** [comment]
@wuminqi I face this issue before it was from `pino.destination` configuration just use `sync: true` and it'll print the errors.

**Comment by AliYusuf95** [comment]
@iamolegga About Skipping the logging in the app startup I have done some workaround to solve this issue.
#### **tl;dt** `app.useLogger` will not work, we have to set the logger when the app created.
As stated in [nestjs docs](https://docs.nestjs.com/techniques/logger#dependency-injection) `The only downside of this solution is that your first initialization messages won't be printed anywhere at all, so in rare cases you may miss some important initialization errors.`

So using this code from the docs will not print all messages as may some expected -myself included-
Yet, it's possible to use `nestjs-pino` as the main logger using [Custom implementation](https://docs.nestjs.com/techniques/logger#custom-implementation) method. 

```ts
import { Logger } from "nestjs-pino";

// Not working
const app = await NestFactory.create(AppModule, { logger: false });
app.useLogger(app.get(Logger));

// Working
const app = await NestFactory.create(ApplicationModule, {
  logger: MyPinoLogger,
});
```

After reviewing how [`PinoLogger`](https://github.com/iamolegga/nestjs-pino/blob/99247e6c64732689c5de8b8dabcaa311bf95c632/src/PinoLogger.ts#L26) works, I found the `outOfContext` logger is assigned only once when the module initialized, then it used for all messages that are executed out of requests context (including Nest App initialization). Which is the same logger used by the [`Logger`](https://github.com/iamolegga/nestjs-pino/blob/99247e6c64732689c5de8b8dabcaa311bf95c632/src/Logger.ts#L7) class.
And for loges executed inside requests context, it uses the one that `pino-http` attached to the request `req.log`. 
So the slution was to create a `PinoLogger` instance before the app created with preconfigured `pino` instance, and I used the same logger instance in `pino-http`.

So this is my implemintation:
```ts
// logger.ts
import { Logger, PinoLogger } from 'nestjs-pino';
import * as pino from 'pino';

const appLogger = pino(options, pino.destination(destOptions));

/**
 * Main Pino child logger used by nestjs app,
 * This logger will be used from any logger injected from LoggerModule
 */
export const MainLogger = appLogger.child({
    level: 'info',
    serializers,
    context: 'LoggerModule',
});

/**
 * Wrap Pino logger into {@link PinoLogger} instance
 */
const pinoLogger = new PinoLogger({
    pinoHttp: {
        logger: MainLogger,
    },
});

/**
 * Create nestjs app logger using Pino logger,
 * This instance needed to use pino logger before instantiate the LoggerModule
 * @see {@link https://docs.nestjs.com/techniques/logger}
 */
export const NestLogger = new Logger(pinoLogger, {});

```
```ts
// main.ts
import { NestLogger } from './logger';

const app = await NestFactory.create(AppModule,{
    logger: NestLogger
});
```
```ts
// app.module.ts
import { MainLogger } from '../logger';

@Module({
    imports: [
        LoggerModule.forRoot({
            pinoHttp: {
                logger: MainLogger,
                reqCustomProps: () => {
                    return { context: 'PinoHttp' };
                },
                genReqId: () => uuidv4(),
            },
            forRoutes: [{
                path: '(.*)',
                method: RequestMethod.ALL,
            }],
        }),
    ],
})
export class AppModule {}
```

**Comment by iamolegga** [comment]
Thanks @AliYusuf95 for your investigation! Main problem here is that if we are using some other nestjs providers for making config for nestjs-pino we should either duplicate the logic for building this config out of nestjs providers for consistency, or use another _default for your logger_ configuration.
In other words problem is: we need to build logger before building app -> we need to build app to get config from providers to build logger

**Comment by AliYusuf95** [comment]
Mainly it based on usage, if it needs to be dynamic with providers then it has to wait for app creation. Otherwise, it can be used independently. 
I think since logging is an essential process that may probably be needed before creating the app -_e.g. logging nest factory_-, it's fine to build the logger before the app. After that, we can extend the logger after app creation if needed -which is exactly the case if we use `app.useLogger(app.get(Logger));`-

**Comment by iamolegga** [comment]
solved in v2.0.0

**Comment by StevenLangbroek** [comment]
Hey @iamolegga! I'm still experiencing this (startup logs are not passed through `pino` and not json) on version 3.1.1 and NestJS 9:

![image](https://user-images.githubusercontent.com/296796/199674697-0be406d7-0715-4025-ba8c-bb826367a98e.png)

My understanding from reading this issue is that it should have been solved? My `main.ts` looks like this:

```ts
  const app = await NestFactory.create<NestExpressApplication>(AppModule);
  const logger = app.get(Logger);
  app.useLogger(logger);
```

I've also tried adding `{ bufferLogs: true }`, but that just results in no logs at all...

What should I do here? Open a new issue? Thanks <3

**Comment by iamolegga** [comment]
@StevenLangbroek please add the link to the minimal example repo. I'll reopen this issue then

**Comment by StevenLangbroek** [comment]
@iamolegga here you go :) https://codesandbox.io/s/async-cookies-m98msq?file=/src/main.ts

**Comment by iamolegga** [comment]
@StevenLangbroek sorry, I'm not familiar with the codesandbox, don't know how to restart server, watch logs, etc, and don't have time to learn it at the moment . Could you create a GITHUB repo please

**Comment by StevenLangbroek** [comment]
༼ つ ◕_◕ ༽つ https://github.com/StevenLangbroek/nestjs-pino-applogger-repro :)

**Comment by iamolegga** [comment]
```ts
  const app = await NestFactory.create(AppModule);
```

```sh
[10:54:39] Starting compilation in watch mode...

[10:54:40] Found 0 errors. Watching for file changes.

[Nest] 64686  - 03.11.2022, 10:54:40     LOG [NestFactory] Starting Nest application...
[Nest] 64686  - 03.11.2022, 10:54:40     LOG [InstanceLoader] AppModule dependencies initialized +17ms
[Nest] 64686  - 03.11.2022, 10:54:40     LOG [InstanceLoader] LoggerModule dependencies initialized +0ms
{"level":30,"time":1667465680411,"pid":64686,"hostname":"macbook-iamolegga.local","context":"RoutesResolver","msg":"AppController {/}:"}
{"level":30,"time":1667465680413,"pid":64686,"hostname":"macbook-iamolegga.local","context":"RouterExplorer","msg":"Mapped {/, GET} route"}
{"level":30,"time":1667465680413,"pid":64686,"hostname":"macbook-iamolegga.local","context":"NestApplication","msg":"Nest application successfully started"}
```

---

```ts
  const app = await NestFactory.create(AppModule, { bufferLogs: true });
```

```sh
[10:55:31] File change detected. Starting incremental compilation...

[10:55:31] Found 0 errors. Watching for file changes.

{"level":30,"time":1667465731578,"pid":64808,"hostname":"macbook-iamolegga.local","context":"NestFactory","msg":"Starting Nest application..."}
{"level":30,"time":1667465731578,"pid":64808,"hostname":"macbook-iamolegga.local","context":"InstanceLoader","msg":"AppModule dependencies initialized"}
{"level":30,"time":1667465731578,"pid":64808,"hostname":"macbook-iamolegga.local","context":"InstanceLoader","msg":"LoggerModule dependencies initialized"}
{"level":30,"time":1667465731578,"pid":64808,"hostname":"macbook-iamolegga.local","context":"RoutesResolver","msg":"AppController {/}:"}
{"level":30,"time":1667465731578,"pid":64808,"hostname":"macbook-iamolegga.local","context":"RouterExplorer","msg":"Mapped {/, GET} route"}
{"level":30,"time":1667465731578,"pid":64808,"hostname":"macbook-iamolegga.local","context":"NestApplication","msg":"Nest application successfully started"}
```

@StevenLangbroek  must be something with your environment

**Comment by StevenLangbroek** [comment]
That's weird 🤔 it happens in Kubernetes pod logs, iTerm and Warp for me... I'll dig more. thanks so far!

**Comment by iamolegga** [comment]
I've checked in iTerm Build 3.4.16. Moreover there is [test](https://github.com/iamolegga/nestjs-pino/blob/master/__tests__/levels.spec.ts#L85-L95) for checking startup logs

**Comment by iamolegga** [comment]
Faced with the same issue today on production (in k8s too), logs were in the same format as [here](https://github.com/iamolegga/nestjs-pino/issues/387#issuecomment-1301774847). After upgrading to the latest versions of nestjs-pino and pino-http it was solved

**Comment by iamolegga** [comment]
Logs in local development were in json format though

**Comment by sezanzeb** [comment]
This seems to work for me, on a first glance

```ts
	const app = await NestFactory.createApplicationContext(
		AppModule,
		{
			logger: new Logger(new PinoLogger({}), {})
		}
	);
```

**Comment by sezanzeb** [comment]
nest-winston also uses that option, see https://www.npmjs.com/package/nest-winston#replacing-the-nest-logger-also-for-bootstrapping

though they seem to provide a special method for convenience for that

**Comment by iamolegga** [comment]
Just upgrade to the latest version and use `{ bufferLogs: true }` as it was already mentioned before. Locking this issue
